### ISSUE TITLE
typetool 3.49.3 problem with `Issue7029Nullness.java`

### ISSUE THREAD

**Issue by wmdietl** [eisop/checker-framework#1213]
CI for #1212 is failing with:

````
NullnessAssumeInitializedTest > run[/home/runner/work/checker-framework/checker-framework/checker/tests/nullness/java17] FAILED
    java.lang.AssertionError: 22 out of 22 expected diagnostics were found.
    2 unexpected diagnostics were found:
      Issue7029Nullness.java:20: error: (return.type.incompatible)
      Issue7029Nullness.java:22: error: (return.type.incompatible)
````

I can locally reproduce the issue:

````
> ./checker/bin/javac -processor nullness ./checker/tests/nullness/java17/Issue7029Nullness.java
Note: The Checker Framework is tested with JDK 8, 11, 17, and 21. You are using version 23.
./checker/tests/nullness/java17/Issue7029Nullness.java:20: error: [return.type.incompatible] incompatible types in return.
              return id == null ? Stream.empty() : Stream.of(Map.entry(id, item));
                                ^
  type of expression: @NonNull Stream<? extends @NonNull Entry<@Nullable String, @NonNull Item>>
  method return type: @Nullable Stream<? extends @NonNull Entry<@NonNull String, @NonNull Item>>
./checker/tests/nullness/java17/Issue7029Nullness.java:22: error: [return.type.incompatible] incompatible types in return.
        .collect(Collectors.toUnmodifiableMap(entry -> entry.getKey(), entry -> entry.getValue()));
                                                                   ^
  type of expression: @Nullable String
  method return type: @NonNull String
2 errors
````

The commit that added that test also modified framework/src/main/java/org/checkerframework/framework/util/typeinference8/util/Resolution.java
That file is the same between the two versions.

- [ ] Is the problem caused by different JDK annotations? eisop is up to date with typetools.
- [ ] Is something about handling records different?
- [ ] What is causing this?

I'm suppressing the warning for now, to get the other improvements in this release.

**Comment by Ao-senXiong** [comment]
Looking into this. I tried transforming `Item` to a Java class and also got the same errors.

**Comment by wmdietl** [comment]
Annoyingly, https://github.com/eisop/checker-framework/commit/9e2c5af915850f68e1846475d9cf164552c85723 typecheck-part2 is failing with:

````
> Task :framework:checkNullness
/home/runner/work/checker-framework/checker-framework/framework/src/main/java/org/checkerframework/framework/type/SubtypeIsSubsetQualifierHierarchy.java:39: warning: [nullness:unneeded.suppression] warning suppression "keyfor:argument.type.incompatible" is not used by NullnessChecker
    @SuppressWarnings({
    ^
/home/runner/work/checker-framework/checker-framework/framework/src/main/java/org/checkerframework/framework/type/SubtypeIsSubsetQualifierHierarchy.java:39: warning: [nullness:unneeded.suppression] warning suppression "nullness:argument.type.incompatible" is not used by NullnessChecker
    @SuppressWarnings({
    ^
error: warnings found and -Werror specified
1 error
2 warnings
````

That commit only contains version updates and the JDK 21 version passed.

**Comment by Ao-senXiong** [comment]
> Annoyingly, [9e2c5af](https://github.com/eisop/checker-framework/commit/9e2c5af915850f68e1846475d9cf164552c85723) typecheck-part2 is failing with:
> 
> ```
> > Task :framework:checkNullness
> /home/runner/work/checker-framework/checker-framework/framework/src/main/java/org/checkerframework/framework/type/SubtypeIsSubsetQualifierHierarchy.java:39: warning: [nullness:unneeded.suppression] warning suppression "keyfor:argument.type.incompatible" is not used by NullnessChecker
>     @SuppressWarnings({
>     ^
> /home/runner/work/checker-framework/checker-framework/framework/src/main/java/org/checkerframework/framework/type/SubtypeIsSubsetQualifierHierarchy.java:39: warning: [nullness:unneeded.suppression] warning suppression "nullness:argument.type.incompatible" is not used by NullnessChecker
>     @SuppressWarnings({
>     ^
> error: warnings found and -Werror specified
> 1 error
> 2 warnings
> ```
> 
> That commit only contains version updates and the JDK 21 version passed.

I am opening https://github.com/eisop/checker-framework/pull/1218 to find out.

**Comment by Ao-senXiong** [comment]
Summary: This issue is due to a type inference bug. In this https://github.com/eisop/jdk/commit/900f1dbddaf509648a319bf71a22e4c461002724, we have removed the annotations on `static <K extends Object, V extends Object> Entry<K, V> entry(K k, V v) `.

If we use `static <K extends Object, V extends Object> Entry<K, V> entry(@NonNull K k, @NonNull V v)`, it works fine
